# Dealing with digits of a number
- 123456
- Find the count of digits in given number
- Find the sum of digits in the given number
- Find the product of the digits in the given number
- Reverse the given number
- Find the sum of squaures of digits of given number
- Find the count of even odd digits separately

# Printing digits of the given number

In [9]:
n = int(input("Enter a number: "))
while n > 0: 
    r = n % 10 
    print(r, end = ' ') 
    n = n // 10

Enter a number:  123456


6 5 4 3 2 1 

# Find the number of digits in the given number (Digit Count)

In [12]:
n = int(input("Enter a number: "))
digit_count = 0
while n > 0: 
    digit_count += 1
    n = n // 10 
print(digit_count)

Enter a number:  123456


6


# Find the sum and product of digits of a given number

In [13]:
n = int(input("Enter a number: "))
s = 0 
p = 1
while n > 0: 
    r = n % 10
    s += r
    p *= r
    n = n // 10 
print(f'Sum is: {s}')
print(f'Product is: {p}')

Enter a number:  1234


Sum is: 10
Product is: 24


# Problem 1
- [Subtract Product and Sum of digits of a number](https://leetcode.com/problems/subtract-the-product-and-sum-of-digits-of-an-integer/)

```java
class Solution {
    public int subtractProductAndSum(int n) {
        int sum = 0;
        int product = 1;
        // Digits 
        while (n > 0) {
            int r = n % 10; // last digit of n
            sum = sum + r; // sum += r;
            product = product * r;
            n = n / 10;
        }
        return product - sum;
    }
}
```

```python
class Solution:
    def subtractProductAndSum(self, n: int) -> int:
        s = 0
        p = 1
        while n > 0:
            r = n % 10 # last digit
            s += r # sum = sum + r
            p *= r # p = p * r
            n = n // 10
        return p - s
```

# Finding Even Odd digit count seperately in the given number

In [40]:
n = int(input('Enter a number: '))
even_digit_count = odd_digit_count = 0
while n > 0:
    r = n % 10
    if r % 2 == 0:
        even_digit_count += 1
    else:
        odd_digit_count += 1
    n = n // 10
print(f'Number of even digits: {even_digit_count}')
print(f'Number of odd digits: {odd_digit_count}')

Enter a number:  123456


Number of even digits: 3
Number of odd digits: 3


# Add Digits
[Add Digits](https://leetcode.com/problems/add-digits/)

- Java Version

```java
class Solution {
    public int getDigitSum(int n) {
        int sum = 0;
        while (n > 0) {
            sum += n % 10;
            n = n / 10;
        }
        return sum;
    }
    public int addDigits(int num) {
        while (num > 9) {
            num = getDigitSum(num);
        }
        return num;
    }
}
```
- Python Version 1

```python
class Solution:
    def addDigits(self, num: int) -> int:
        s = 0
        while num > 0: 
            r = num % 10
            s += r # s = 11
            num = num // 10 # n = 11
            if num == 0 and s >= 10:
                num = s
                s = 0
        return s
```
- Python Version 2

```python
class Solution:
    def addDigits(self, num: int) -> int:
        if num % 9 == 0 and num != 0:
            return 9
        return num % 9
```

# Add digits repeatedly

In [15]:
def add_digits(n):
    s = 0
    while n > 0: 
        r = n % 10
        s += r
        n = n // 10 
    return s
    
n = 38
while n > 10:
    n = add_digits(n)
print(n)

2


In [16]:
n = 38
s = 0
while n > 0: 
    r = n % 10
    s += r # s = 11
    n = n // 10 # n = 11
    if n == 0 and s >= 10:
        n = s
        s = 0
print(s)

2


# Reversing a number

In [26]:
n = int(input('Enter a number: '))
rev = 0
while n > 0:
    r = n % 10
    rev = rev * 10 + r
    n = n // 10
print(rev)

Enter a number:  123


321


# Checking if the given number is a palindrome
- 123 is not a palindrome, because reverse  of 123 is 321 which is not equal to 123
- 242 is a palindrome, because reverse of 242 is also 242

In [25]:
n = int(input('Enter a number: '))
rev = 0
t = n
while n > 0:
    r = n % 10
    rev = rev * 10 + r
    n = n // 10
if rev == t:
    print(f'{t} is a palindrome')
else:
    print(f'{t} is not a palindrome')

Enter a number:  121


121 is a palindrome


# A Number after double reverse
[A Number after double reverse](https://leetcode.com/problems/a-number-after-a-double-reversal/description/)

- Java Version
```java
class Solution {
    public int getReverse(int n) {
        int rev = 0;
        while (n > 0) {
            int r = n % 10;
            rev = rev * 10 + r;
            n = n / 10;
        }
        return rev;
    }
    public boolean isSameAfterReversals(int num) {
        int rev1 = getReverse(num);
        int rev2 = getReverse(rev1);
        return rev2 == num;
    }
}
```

- Python Version

```python
class Solution:
    def get_reverse(n):
        rev = 0
        while n > 0:
            rev = rev * 10 + (n % 10)
            n = n // 10
        return rev
    def isSameAfterReversals(self, num: int) -> bool:
       rev1 = Solution.get_reverse(num)
       rev2 = Solution.get_reverse(rev1)
       return rev2 == num
```

# Happy Number

In [ ]:
def get_square_sum(n):
    sq_sum = 0
    while n > 0:
        r = n % 10
        sq_sum += r * r # square sum of digit
        n = n // 10
    return sq_sum

n = 2
# Finding square sum repeatedly until we hit a single digit
while n > 9:
    n = get_square_sum(n)
if n == 1 or n == 7:
    print('Happy')
else:
    print('Not Happy')

# Happy Number
[Happy Number](https://leetcode.com/problems/happy-number/description/)

- Java Version
```java
class Solution {
    public int getSquareSum(int n) {
        int sqSum = 0;
        while (n > 0) {
            int r = n % 10;
            sqSum += r * r;
            n = n / 10;
        }
        return sqSum;
    }
    public boolean isHappy(int n) {
        while (n > 9) // until single digit
        {
            n = getSquareSum(n);
        }
        return (n == 1 || n == 7);
    }
}
```

- Python Version
```python
class Solution:
    def get_square_sum(n: int) -> int:
        sq_sum = 0
        while n > 0:
            r = n % 10
            sq_sum += r * r
            n = n // 10
        return sq_sum

    def isHappy(self, n: int) -> bool:
        while n > 9:
            n = Solution.get_square_sum(n)
        return n == 1 or n == 7
```